# WhisperX for uploading files

Upload your local files to the Colab Files from the left sidebar.

从左侧将视频音频文件上传到Colab，然后运行即可下载生成好的字幕文件。



In [1]:
# @title *通用参数/Required settings:**
# Check GPU availability
!nvidia-smi

# @markdown #### **Initial prompt**
# @markdown Prompts can be very helpful for correcting specific words or acronyms that the model often misrecognizes in the audio.
prompt = "ChatGPT, LLM, DALL-E,Turbo"  # @param {type:"string"}

# @markdown #### **Directory Path**
# @markdown where your audio-video files are located from Coloab
directory_path = "/content"  # @param {type:"string"}

# @markdown #### Model
model_size = "large-v3"  # @param ["base", "base.en", "small", "small.en","medium", "medium.en", "large-v1","large-v2","large-v3"]

# @markdown #### Language
language = "auto" # @param ["auto", "en", "zh", "de", "es", "ru", "ko", "fr", "ja", "pt", "tr", "pl", "ca", "nl", "ar", "sv", "it", "id", "hi", "fi", "vi", "he", "uk", "el", "ms", "cs", "ro", "da", "hu", "ta", "no", "th", "ur", "hr", "bg", "lt", "la", "mi", "ml", "cy", "sk", "te", "fa", "lv", "bn", "sr", "az", "sl", "kn", "et", "mk", "br", "eu", "is", "hy", "ne", "mn", "bs", "kk", "sq", "sw", "gl", "mr", "pa", "si", "km", "sn", "yo", "so", "af", "oc", "ka", "be", "tg", "sd", "gu", "am", "yi", "lo", "uz", "fo", "ht", "ps", "tk", "nn", "mt", "sa", "lb", "my", "bo", "tl", "mg", "as", "tt", "haw", "ln", "ha", "ba", "jw", "su"]


# @markdown #### Assign speaker labels
# @markdown Recognize speakers
assign_speaker_lable = False # @param {type:"boolean"}

# @markdown #### Align whisper output
align_whisper_output = True # @param {type:"boolean"}

Mon Mar  3 05:49:31 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   54C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

conda install pytorch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1 pytorch-cuda=12.4 -c pytorch -c nvidia# Install WhisperX

In [2]:
! pip install torch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1 --index-url https://download.pytorch.org/whl/cu124

! pip install ctranslate2==4.5.0
! pip install git+https://github.com/m-bain/whisperx.git

# ! pip install git+https://github.com/openai/whisper.git


# !pip install git+https://github.com/federicotorrielli/BetterWhisperX

# !pip install git+https://github.com/Hasan-Naseer/whisperX.git@release/latest-faster-whisper-version



/bin/bash: line 1: conda: command not found
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 47.6 MB/s eta 0:00:00
  Cloning https://github.com/m-bain/whisperx.git to /tmp/pip-req-build-ula_1ve7
  Running command git clone --filter=blob:none --quiet https://github.com/m-bain/whisperx.git /tmp/pip-req-build-ula_1ve7
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/pip/_internal/cli/base_command.py", line 179, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/pip/_internal/cli/req_command.py", line 67, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/pip/_internal/commands/install.py", line 377, in run
    requirement_set = resolver.resolve(
                      ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/pip/_internal/resolution/resolvelib/resolver.py", line 76, 

In [1]:
# Please BE VERY CAREFUL, this will link your entire drive.
# So don't edit code, except the one that says 'Customize the following options',
# or you might mess up your files.
# IF YOU DO NO WANT TO LINK DRIVE, please see below for an alternative!
from google.colab import drive
drive.mount('/gdrive')

Mounted at /gdrive


# Get Subtitle Files using WhisperX

In [4]:

import os
from google.colab import files

in_path = '/gdrive/MyDrive/media_proc/speedreg/in/'
out_path = '/gdrive/MyDrive/media_proc/speedreg/out/'
directory_path = in_path

# supported extensions
supported_extensions = ['.mp4', '.wav', '.mp3']

language_param = ""
if language != "auto":
    language_param = f"--language {language}"

diarize_param = ""
if assign_speaker_lable:
    diarize_param = "--diarize --hf_token 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"

align_whisper_param = ""
if align_whisper_output:
    align_whisper_param = "--align_model WAV2VEC2_ASR_LARGE_LV60K_960H"

prompt_param = ""
if prompt != "":
    prompt_param = f'--initial_prompt "{prompt}"'

# def process_file(filename):
#     # run = f'whisperx "/content/APO2992689654.mp3" --max_line_count 1 --max_line_width 100 --model medium.en --diarize --hf_token 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 --output_dir . --align_model WAV2VEC2_ASR_LARGE_LV60K_960H'
#     run = f'whisperx "{filename}" --model {model_size} {language_param} --output_dir . {prompt_param} {diarize_param} {align_whisper_param}'
#     !{run}

#     print("Start to download subtitle files")
#     # start to download file
#     base_filename = os.path.splitext(filename)[0]
#     srt_filename = base_filename + '.srt'
#     json_filename = base_filename + '.json'

#     files.download(srt_filename)
#     files.download(json_filename)

import os, shutil
def calculate_relative_path(directory_path, filename):
    # 获取指定目录的绝对路径
    abs_directory_path = os.path.abspath(directory_path)

    # 获取给定文件名的绝对路径
    abs_file_path = os.path.abspath(filename)

    # 确保给定的文件名在指定目录下的子目录树中
    if not abs_file_path.startswith(abs_directory_path):
        raise ValueError("The given filename is not in the specified directory tree.")

    # 计算相对路径
    relative_path = os.path.relpath(abs_file_path, abs_directory_path)
    return relative_path

def make_out_relative_path(out_path, filename):
    # 计算出filename的路径在directory_path下的子目录树
    relative_path = calculate_relative_path(in_path, filename)
    print(relative_path)
    existing_directory = os.path.join(out_path, relative_path)
    if os.path.exists(existing_directory):
        shutil.rmtree(existing_directory)

    # 在另一个目录下创建子目录
    new_directory = os.path.join(out_path, relative_path)
    os.makedirs(new_directory, exist_ok=True)
    return new_directory

def process_file(filename):

    new_directory = make_out_relative_path(out_path, filename)
    print(new_directory)
    # run = f'whisper_timestamped "/content/APO2992689654.mp3" --max_line_count 1 --max_line_width 100 --model medium.en --diarize --hf_token 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 --output_dir . --align_model WAV2VEC2_ASR_LARGE_LV60K_960H'
    # run = f'whisper_timestamped "{filename}" --model {model_size} {language_param} --output_dir {new_directory} {prompt_param} {diarize_param} {align_whisper_param}'
    # !{run}

    run = f'whisperx "{filename}" --model {model_size} {language_param} --output_dir {new_directory} {prompt_param} {diarize_param} {align_whisper_param}'
    !{run}

    # print("Start to download subtitle files")
    # # start to download file
    # base_filename = os.path.splitext(filename)[0]
    # srt_filename = base_filename + '.srt'
    # json_filename = base_filename + '.words.json'

    # files.download(srt_filename)
    # files.download(json_filename)

def process():
    for root, dirs, files in os.walk(directory_path):
        for file in files:
            if file.endswith(tuple(supported_extensions)):
                file_path = os.path.join(root, file)
                print(file_path)
                process_file(file_path)

process()

/gdrive/MyDrive/media_proc/speedreg/in/DiaryofaWimpyKid16.mp3
DiaryofaWimpyKid16.mp3
/gdrive/MyDrive/media_proc/speedreg/out/DiaryofaWimpyKid16.mp3
2025-03-03 05:46:10.676299: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1740980770.975008    1485 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1740980771.048566    1485 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-03-03 05:46:11.622361: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appro

In [8]:
!find / -name libcudnn_ops_infer.so.8

find: ‘/proc/68/task/68/net’: Invalid argument
find: ‘/proc/68/net’: Invalid argument
find: ‘/proc/2766/task/2766/net’: Invalid argument
find: ‘/proc/2766/net’: Invalid argument


In [10]:
!find . -name 'libcublas.so*' -o -name 'libcudnn_ops*'